pkg/ebpf/c/tracee.bpf.c

```diff
@@ -5153,73 +5182,175 @@ CGROUP_SKB_HANDLE_FUNCTION(proto_icmpv6);
+    u64 flags = BPF_F_CURRENT_CPU | (u64) size << 32;
     neteventctx->bytes = size;
 
-    // set the event type before submitting event
+    // Set the event type before submitting event.
     neteventctx->eventctx.eventid = event_type;
 
-    return bpf_perf_event_output(ctx,
-                                 map,
-                                 flags,
-                                 neteventctx,
-                                 sizeof_net_event_context_t());
+    // Submit the event.
+    return bpf_perf_event_output(ctx, map, flags, neteventctx, sizeof_net_event_context_t());
 }
 
-#define cgroup_skb_submit_event(a,b,c,d) cgroup_skb_submit(&events,a,b,c,d)
+// Submit a network event.
+#define cgroup_skb_submit_event(a, b, c, d) cgroup_skb_submit(&events, a, b, c, d)
+
+// Check if a flag is set in the retval.
+#define retval_hasflag(flag) (neteventctx->eventctx.retval & flag) == flag
+
+// Keep track of a flow event if they are enabled and if any policy matched.
+// Submit the flow base event so userland can derive the flow events.
+statfunc u32 cgroup_skb_submit_flow(struct __sk_buff *ctx,
+                                    net_event_context_t *neteventctx,
+                                    u32 event_type, u32 size, u32 flow)
+{
+    netflowvalue_t *netflowvalptr, netflowvalue = {
+                                       .last_update = bpf_ktime_get_ns(),
+                                       .direction = flow_unknown,
+                                   };
+
+    // Set the current netctx task as the flow task.
+    neteventctx->md.flow.host_pid = neteventctx->eventctx.task.host_pid;
+
+    // Set the flow event type in retval.
+    neteventctx->eventctx.retval |= flow;
+
+    // Check if the current packet source is the flow initiator.
+    bool is_initiator = 0;
+
+    switch (flow) {
+        // 1) TCP connection is being established.
+        case flow_tcp_begin:
+            // Ingress: Remote (src) is sending SYN+ACK: this host (dst) is the initiator.
+            if (retval_hasflag(packet_ingress))
+                netflowvalue.direction = flow_outgoing;
+
+            // Egress: Host (src) is sending SYN+ACK: remote (dst) host is the initiator.
+            if (retval_hasflag(packet_egress))
+                netflowvalue.direction = flow_incoming;
+
+            // Invert src/dst: The flowmap src should always be set to flow initiator.
+            neteventctx->md.flow = invert_netflow(neteventctx->md.flow);
+
+            // Update the flow map.
+            bpf_map_update_elem(&netflowmap, &neteventctx->md.flow, &netflowvalue, BPF_NOEXIST);
+
+            break;
+
+        // 2) TCP connection is being closed/terminated.
+        case flow_tcp_end:
+            // Any side can close the connection (FIN, RST, etc). Need heuristics.
+
+            // Attempt 01: Try to find the flow using current src/dst.
+
+            for (int n = 0; n < 3; n++) {
```

Comment:
The problem this loop tries to solve is the following, ingress usually run from "idle", "ksoftirqd" or "a process" (all in SOFTIRQ context), with higher privilege.

IF there is a bpf map update from egress, bpf_map_update() returned but the value hasn't been fully synched (among different CPU caches, with a barrier, or bpf program preemption, as we've seen before), and the connection is too fast, the ingress FIN packet would execute the bpf program with a higher privilege (perhaps in other CPU), and the map wouldn't contain the entry.

This loop gives at least a bit of time for things to settle (and there might be a barrier/synch for the map, or not). In my tests it was "enough".

Reply:
I did get the issue in general it's trying to solve, I was just not sure if the given backoff is enough. Maybe this should be configurable. I don't know if there is something else we can add.

Reply:
Not in the eBPF program context I'm afraid. In kernel, I would be SURE if the bpf map value ptr (thus the values) has changed, or not, I suppose (I can always force a barrier). This way, readers would know it atomically. 

In eBPF ISA I have atomic instructions only for map values that already exist, for example, but when it comes to the entry in the map either "existing or not", its tricky. That because calling the bpf helper (from eBPF VM into kernel) so it can update the map value doesn't guarantee no preemptions afaik.

So, its a "best effort" approach, in summary.